### PR TITLE
Introducing Region as record for supporting non-destructive mutations

### PR DESCRIPTION
**Pull request by danigian**
# PR for issue #1082 

## What is being addressed

Region is now a record and all its properties are only settable at initialization. In the case of AS923, the WithFrequencyOffset is creating a copy of the region record for a non-destructive mutation of the frequencyoffset field

### CODE REVIEW

File: Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs [hunk 1]
```diff
@@ -580,5 +580,110 @@ public void RegionConfigurationConverter_Throws_OnNotSetRegion()
...
+            'nodwell': true}}";
+
+            var config1 = JsonUtil.Strictify(string.Format(CultureInfo.InvariantCulture, input, 923000000, 922000000, 200000, 400000));
+            var region1 = LnsStationConfiguration.GetRegion(config1);
+
+            var config2 = JsonUtil.Strictify(string.Format(CultureInfo.InvariantCulture, input, 923000000, 922000000, -1600000, -1400000));
+            var region2 = LnsStationConfiguration.GetRegion(config2);
+
+            // asserting that reading region2 is not affecting region1
+            Assert.Equal(typeof(RegionAS923), region1.GetType());
```
Comment (bastbu) [inline Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs:683]: Consider using `Assert.IsType` and leveraging the return value to continue with further assertions (without needing to cast)

File: Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs [hunk 2]
```diff
@@ -580,5 +580,110 @@ public void RegionConfigurationConverter_Throws_OnNotSetRegion()
...
+                     [ 9, 125, 0 ],
+                     [ 8, 125, 0 ],
+                     [ 7, 125, 0 ],
+                     [ 7, 250, 0 ] ],
+            'sx1301_conf': [
+                {{
+                    'radio_0': {{
+                        'enable': true,
+                        'type': 'SX1257',
+                        'freq': {0},
```
Comment (bastbu) [inline Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs:605]: [nit] we might want to think about integrating this into the helper methods that we already have to replace tokens in that template. In that case, if the format changes, we don't have to fix it all over the place but just in a single place.

**Review by bastbu — APPROVED** [review]

File: Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs (hunk 1 above)
Comment (danigian) [reply Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs]: Done

File: Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs (hunk 2 above)
Comment (danigian) [reply Tests/Unit/NetworkServer/JsonHandlers/LnsStationConfigurationTests.cs:605]: Even though I agree with you, I think we should not do this as part of this PR.

Could you please discuss the idea tomorrow in the daily and eventually create a new issue, @bastbu? Thanks 👍

**Review by atifaziz — COMMENTED** [review]
While I see the rationale/motivation, I think it's dangerous to make `Region` a record if the other objects it holds like `RX1DROffsetTable` are mutable. The object is only partially immutable and might lead to subtle bugs.

**Comment by danigian** [comment]
After discussing with @atifaziz regarding his comment on underlying mutable properties, I decided to put back this PR in draft.

Currently Dictionaries like 'DRtoConfiguration' are exposed as public properties. We should either make this Dictionary readonly collections or we should make them not public and only provide methods for accessing what needs to be accessed.

**Comment by danigian** [comment]
#1119 should fix the issue related to this pr, therefore i am closing this one

**Pull request closed without merging**